doc/experimental_ui_api.md

```diff
@@ -0,0 +1,2027 @@
+  element. This also matches the root element.
+* `?nth_child(index)` matches all elements that are at the `index` position
+  from the start of their parent. So, `?nth_child(3)` matches the third child
+  of each element.
+* `?nth_last_child(index)` matches all elements that are at the `index`
+  position from the end of their parent.
+* `?<(sel)` matches all elements whose parent is matched by the selector `sel`.
+* `?>(sel)` matches all elements that have a child that is matched by the
+  selector `sel`.
+* `?<<(sel)` matches all elements that have an ancestor that is matched by the
+  selector `sel`.
+* `?>>(sel)` matches all elements that have a descendant that is matched by the
+  selector `sel`.
+* `?<>(sel)` matches all elements that have a sibling that is matched by the
+  selector `sel`.
+* `?family(name)` matches all elements that have a family of `name`. If `name`
+  is `*`, then it matches any elements that have any family at all.
+
+Unlike CSS, there are no child or descendant combinators since the more
+powerful `?<()` and `?<<()` predicates can be used instead. Note that
+predicates like `?>>()` should be used with care since they may have to check a
+large number of elements in order to see if there is a match, which may cause
+loss of performance for particularly large UIs.
+
+`StyleSpec`
+------------
+
+A `StyleSpec` is a plain table of properties for use in `ui.Style` or as
+inlined properties in `ui.Elem`.
+
+### Field formats
+
+`StyleSpec` has specific field formats for positions and rectangles:
+
+* 2D vector: A table `{x, y}` representing a position, size, or offset. The
+  shorthand `{num}` can be used instead of `{num, num}`.
+* Rectangle: A table `{left, top, right, bottom}` representing a rectangle or
+  set of borders. The shorthand `{x, y}` can be used instead of `{x, y, x, y}`,
+  and `{num}` can be used instead of `{num, num, num, num}`.
+
+### Fields
+
+All properties are optional. Invalid properties are ignored.
+
+#### Layout fields
+
+* `layout` (string): Chooses what layout scheme this box will use when laying
+  its children out. Currently, the only valid option is `place`.
+* `clip` (string): Normally, a box expands its minimum size if there's not
+  enough space for the content, but this property can specify that the content
+  be clipped in the horizontal and/or vertical directions instead. One of
+  `none`, `x`, `y`, or `both`. Default `none`.
+* `scale` (number): The scale factor by which positions and sizes will be
+  multiplied by in `place` layouts.
+    * A scale of zero will scale coordinates by the width and height of the
+      parent box, effectively creating normalized coordinates.
+
+#### Sizing fields
+
+* `size` (2D vector): The minimum size of the box in pixels. May not be
+  negative. Default `{0, 0}`.
+* `span` (2D vector): The size that the layout scheme will allocate for the
+  box, scaled by `scale`. Default `{1, 1}`.
```

Comment:
I see you've done some work to address this:

> `size` is broken by default because `rel_size` is {1, 1} by default. Setting `size` simply does nothing by default. You manually need to set `rel_size = {0, 0}` for it to work. This is confusing.

That's great. However, now we have

- `size`, which is actually a minimum size, always in pixels.
- `span`, which is the actual size, either in pixels or in scaled pixels or in normalized coordinates, depending on `scale`.

Since `span` is `{1, 1}` by default, the problem I reported earlier might even still exist (?) (depends on the (undocumented) default value of `scale`)

And: This is still confusing. Why isn't `span` called `size`? You describe it as "The size [...]".

`span` could be called `min_size`, you describe it as "The minimum size [...]".

Reply:
The idea is that elements should, by default, use relative positioning. The `scale` is 0 by default, so elements will take up their entire parent. I expect that absolute positioning will be somewhat somewhat rarely used once we have sizers.

Nevertheless, choosing absolute positioning is as easy as changing `layout = "flex"`. Since `layout = "place"` by default, you just have to set `scale = 1` or `scale = ui.get_coord_size()` or so on. Then you can use the `span` property just like you'd use it for relative layout.

The reason I used `size` instead of `min_size` is that minimum sizes are rather common to set when using sizers. The `span` property is unused for flex sizers, and it means the number of rows and columns an element spans in a grid sizer. So, I thought that giving it the name `size` would be misleading in conjunction with `min_size`, since it's nothing like CSS `width` vs. `min-width` except in place sizers.

Reply:
I considered units as a distinct possibility, but I worry that they might complicate flex and grid sizers greatly. For instance, if you have an element with a minimum width of `50%` in a flex sizer, then the minimum size of the sizer element depends on the minimum size of the elements, which depend on the size of the sizer. I don't know how CSS resolves such circular dependencies in terms of units.

(There's also the inconvenience of being forced to use strings for units, as `size = {"50%", "100%"}` is cumbersome.)

I like to think of the sizers in the UI API as being a step above some of the fancy sizers in desktop GUI toolkits like Qt or wxWidgets, but a definite step in complexity lower than CSS. Hence why the only thing common to all sizers is the `size` property in pixels, with everything else being up to the sizer and its style properties. `anchor`, for instance, is specific to the place sizer and does nothing for flex or grid sizers.

Minimum sizes for wrapping flex sizers are already a nightmare, BTW, since they have a range of minimum sizes depending on where the sizer wraps, which is why I held off on implementing minimum sizes for so long. I haven't found any completely satisfactory solutions so far. Non-wrapping sizers and grid sizers are more well-behaved.

Reply:
> The reason I used `size` instead of `min_size` is that minimum sizes are rather common to set when using sizers. The `span` property is unused for flex sizers, and it means the number of rows and columns an element spans in a grid sizer. So, I thought that giving it the name `size` would be misleading in conjunction with `min_size`, since it's nothing like CSS `width` vs. `min-width` except in place sizers.

Compromise suggestion: If `span`'s meaning is not consistently `size`, it can keep it's name. Since `size`'s meaning is consistently `min_size`, should still rename it to `min_size`. This makes sense and fixes the "`size` doesn't work by default" confusing behavior.

Better suggestion: That one property (`span`) has an entirely different meaning depending on the sizer used is actually a problem in itself and the root cause of this naming problem. For grid sizers, there should be a `span` property or similar, while for place sizers, there should be a `size` property.